src/liballoc/alloc.rs

```diff
@@ -279,10 +286,12 @@ unsafe fn exchange_malloc(size: usize, align: usize) -> *mut u8 {
 // For example if `Box` is changed to  `struct Box<T: ?Sized, A: AllocRef>(Unique<T>, A)`,
 // this function has to be changed to `fn box_free<T: ?Sized, A: AllocRef>(Unique<T>, A)` as well.
 pub(crate) unsafe fn box_free<T: ?Sized>(ptr: Unique<T>) {
-    let size = size_of_val(ptr.as_ref());
-    let align = min_align_of_val(ptr.as_ref());
-    let layout = Layout::from_size_align_unchecked(size, align);
-    Global.dealloc(ptr.cast().into(), layout)
+    unsafe {
+        let size = size_of_val(ptr.as_ref());
+        let align = min_align_of_val(ptr.as_ref());
+        let layout = Layout::from_size_align_unchecked(size, align);
+        Global.dealloc(ptr.cast().into(), layout)
```

Comment:
this is probably the only unsafe thing? or else this + the last two lines

Reply:
Pretty sure that `ptr.as_ref()` here is unsafe because it's a Unique<T> which doesn't guarantee validity? I could be wrong though.

(FWIW this is part of why I think it may be worth trying to optimize for unsafe scoping to just one op)

Reply:
ah maybe so

Reply:
`ptr.as_ref()` is indeed unsafe, but `size_of_val()`, `min_align_of_val()` and `Layout::from_size_align_unchecked` are unsafe too.